### ISSUE TITLE
Error when an external component changes nodes order (for example JQuery UI Sortable)

### ISSUE THREAD

**Issue by cervengoc** [brianmhunt/knockout-fast-foreach#31]
Because of the way the last node references are maintained, when an external component changes node order it will fail drastically due to the loop in delete function.

Proposed fix is to maintain the first nodes as well (which wouldn't be too much overhead) and rewrite the delete function core to use it instead of pure indexes.

Of curse the rendered items and the bound array will still stay out of sync (which is IMO out of the scope of this project), but importantly the binding won't crash.

Please see the corresponding commit, and review it.

**Comment by brianmhunt** [comment]
Thanks @cervengoc ... When I made this plugin, I called this "DOM state integrity", and had considered exactly this problem.

At first I erred on the side of memory performance for very large arrays, which is slightly advantaged with the current implementation, but on reflection (and your commit) I think maintaining stateful integrity is the preferable route – particularly since it lets one change item order as with jQuery UI.

In other words, I like it. :smile:  Let me review.

**Comment by cervengoc** [comment]
While I was walking home today I was thinking about it, and while this fix can probably be still a good thing, I'm about to come to the conclusion that handling jQuery UI's sortable and such external plugins should be completely out of scope, and handled outside. I will try to experiment with it and report back.

My main reason is that it's okay not to crash at DOM reorder, but still, it's not a good behaviour to let messing up things silently. I mean the array will be still out of sync, and whenever I change the array, it will be even worse. So probably we shouldn't even let this happen and throw an exception when DOM integrity gets corrupted... I don't know, just flushed my thoughts.

As for the sortable plugin, my alternative idea is to handle the event before the item gets dropped in its new place, cancel it, make the corresponding array change based on the event data and using ko.dataFor, and friends. And finally let fastForEach to reorder the visual elements. I think it should not add so much overhead, so users probably couldn't even notice the little "stuttering", What do you think?

**Comment by cervengoc** [comment]
I've added a new commit please check that out.

**Comment by brianmhunt** [comment]
Thanks @cervengoc – While in some ways there's a purity to having knockout (or `knockout-transformations`, etc.) handle the sorting, I don't like it crashing for any reason. :)

I can't see a diff to the new commit; would you mind opening a PR so I can track changes?  Cheers
